Make smtp_tls_policy_maps easily configurable

nextgens opened the pull request:
## What type of PR?

Feature

## What does this PR do?

- Make smtp_tls_policy_maps easily configurable. This is useful to force TLS verification of specific destinations (or relays).
We should probably discuss what's on the list by default. I have found a top100 list online, ran it through a script to check all the records and found 90 destinations we could use.
- disable TLS session tickets (this reduces the PFS window from 1day to 1h)
- enable system CAs by default (to allow for OUTBOUND_TLS_LEVEL above encrypt without additional overrides)

### Related issue(s)
- closes #1558 
- #707

## Prerequistes
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [x] In case of feature or enhancement: documentation updated accordingly
- [x] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/workflow.html#changelog) entry file.

nextgens commented:
bors try

nextgens commented:
bors retry

Diman0 commented:
During the last project meeting, you told us you could make a script available you used for testing?
Can you share it?

When testing I receive errors. I set the ```OUTBOUND_TLS_LEVEL=secure```.
After that I send emails to https://havedane.net/

I can see the following in the logs. Do you see the same when you test the PR?

```
smtp_1      | Aug 19 17:10:45 test postfix/smtpd[369]: disconnect from mailu_webmail_1.mailu_default[192.168.203.8] ehlo=2 xclient=0/1 mail=1 rcpt=3 data=1 quit=1 commands=8/9
smtp_1      | Aug 19 17:10:46 test postfix/smtp[372]: error: unsupported dictionary type: hash
smtp_1      | Aug 19 17:10:46 test postfix/smtp[374]: error: unsupported dictionary type: hash
smtp_1      | Aug 19 17:10:46 test postfix/smtp[375]: error: unsupported dictionary type: hash
smtp_1      | Aug 19 17:10:46 test postfix/smtp[374]: warning: hash:/etc/postfix/tls_policy.map is unavailable. unsupported dictionary type: hash
smtp_1      | Aug 19 17:10:46 test postfix/smtp[374]: warning: hash:/etc/postfix/tls_policy.map lookup error for "dont.havedane.net"
smtp_1      | Aug 19 17:10:46 test postfix/smtp[374]: warning: smtp_tls_policy_maps, next-hop destination "dont.havedane.net": policy table lookup error
smtp_1      | Aug 19 17:10:46 test postfix/smtp[374]: warning: TLS policy lookup for dont.havedane.net/dont.havedane.net: client TLS configuration problem
smtp_1      | Aug 19 17:10:46 test postfix/smtp[374]: warning: TLS policy lookup for dont.havedane.net/dont.havedane.net: client TLS configuration problem
smtp_1      | Aug 19 17:10:46 test postfix/smtp[372]: warning: hash:/etc/postfix/tls_policy.map is unavailable. unsupported dictionary type: hash
smtp_1      | Aug 19 17:10:46 test postfix/smtp[372]: warning: hash:/etc/postfix/tls_policy.map lookup error for "do.havedane.net"
smtp_1      | Aug 19 17:10:46 test postfix/smtp[372]: warning: smtp_tls_policy_maps, next-hop destination "do.havedane.net": policy table lookup error
smtp_1      | Aug 19 17:10:46 test postfix/smtp[372]: warning: TLS policy lookup for do.havedane.net/do.havedane.net: client TLS configuration problem
smtp_1      | Aug 19 17:10:46 test postfix/smtp[372]: warning: TLS policy lookup for do.havedane.net/do.havedane.net: client TLS configuration problem

```

Diman0 commented on core/postfix/start.py:
```diff
@@ -66,6 +66,12 @@ def is_valid_postconf_line(line):
     os.system("postmap {}".format(destination))
     os.remove(destination)
 
+if not os.path.exists("/etc/postfix/tls_policy.map.db"):
```
When testing I see that postfix creates the file tls_policy.map.lmdb instead of tls_policy.map.db.

Diman0 commented on core/postfix/start.py:
```diff
@@ -66,6 +66,12 @@ def is_valid_postconf_line(line):
+    with open("/etc/postfix/tls_policy.map", "w") as f:
+        for domain in ['gmail.com', 'yahoo.com', 'hotmail.com', 'aol.com', 'outlook.com', 'comcast.net', 'icloud.com', 'msn.com', 'hotmail.co.uk', 'live.com', 'yahoo.co.in', 'me.com', 'mail.ru', 'cox.net', 'yahoo.co.uk', 'verizon.net', 'ymail.com', 'hotmail.it', 'kw.com', 'yahoo.com.tw', 'mac.com', 'live.se', 'live.nl', 'yahoo.com.br', 'googlemail.com', 'libero.it', 'web.de', 'allstate.com', 'bti ...
+            f.write(f'{domain}\tsecure\n')
+    os.system("postmap /etc/postfix/tls_policy.map")
```
I think this should be ```postmap hash:/etc/postfix/tls_policy.map``` to create a hash db instead of a  lmdb database.

Diman0 reviewed (changes requested):
See in-line comments. I will check that my comments are correct, but it looks like you create a lmdb db instead of hash db because you did not specify the type in the postmap command.

Diman0 commented:
wait. alpine removed support for hash. See this PR https://github.com/Mailu/Mailu/pull/1918
So in this PR you should change from hash: to lmdb in the main.cf file.

Diman0 commented:
It should now be fine. 

bors try

nextgens commented:
Yes sorry... this PR predates #1918 and I didn't realize that it required the same changes...

nextgens commented:
> During the last project meeting, you told us you could make a script available you used for testing?
> Can you share it?
> 

Sure; here it is; ensure you run it from a host with a valid rDNS record.
```
#!/usr/bin/env python3
#
# check whether all the domains in the list have MX servers
#that can do STARTTLS with valid certificates
#
# I've used the top100 list from https://www.gmass.co/domains
#
# nextgens ~ 2021

from smtplib import SMTP
import ssl
import dns.resolver
import sys

sslctx = ssl.create_default_context()
sslctx.check_hostname = True
sslctx.verify_mode = ssl.CERT_REQUIRED

with open("list.txt","r") as f:
    with open("ok.txt","w") as ok:
            with open("err.txt","w") as err:
                for domain in f:
                    domain = domain.strip()
                    try:
                        for mx in dns.resolver.query(domain, 'MX'):
                            mx_host = mx.to_text().split(' ')[1][:-1]

                            with SMTP(mx_host) as smtp:
                                #smtp.set_debuglevel(2)
                                smtp.ehlo()
                                smtp.starttls(context=sslctx)
                                smtp.ehlo()
                        ok.write(f'{domain}\tsecure\n')
                    except Exception as e:
                            err.write(f'{domain}{e}\n')
```

> When testing I receive errors. I set the `OUTBOUND_TLS_LEVEL=secure`.
> After that I send emails to https://havedane.net/
> 
> I can see the following in the logs. Do you see the same when you test the PR?
> 

You've now fixed the issue (was alpine removing support for BDB)... I wouldn't anticipate anything to work when you are sending mails to @havedane.net though: this PR is about web/PKIX support (hosts that have a certificate signed by a recognized CA), not DANE.

It does make configuring DANE for select destinations easier (by providing an easy to override tls_policy.map) but won't do it for you.

Diman0 commented:
Thank you for providing the script. I will do some testing when I have time (possibly today) and will then approve the PR.

nextgens commented:
To put it in simple words: postfix as configured by mailu without the PR will do opportunistic encryption: it will use TLS to provide protection against a passive attacker.

With this PR, it will also provide protection against an active attacker (by verifying/validating the TLS certificate) for select domains.

We should probably discuss what's worth having on the list: it could be much bigger; the only thing we need to trust is that they won't fail at maintaining valid TLS certificates going forward.

nextgens commented:
bors try

Diman0 reviewed (approved):
Lgtm
